### ISSUE TITLE
sql: [strawman] simplify date/time/timestamp handling by ignoring session offset

### ISSUE THREAD

**Issue by dt** [cockroachdb/cockroach#9596]
[moving an offline discussion of "is this even worth spending time on?" here for broader participation]

## TL;DR
Disallow non-zero session zone offsets, always render timestamps in UTC and with zero offset.
## Summary
Session time zone offsets are very subtle and potentially confusing -- when and where they are (or should be) used or not has proven unintuitive and consumed a disproportionate amount of time and energy. Rendering times in a particular offset (which is basically all it is used for) is primarily a application/presentation layer concern, and we might be best off leaving it to applications and focusing elsewhere for now.

## Background
The SQL spec (and postgres) says that a `timestamp` is the same as a `timestamp without time zone`. The difference between `with time zone` and `without time zone` timestamps is how they are rendered to the client -- `with time zone` is rendered _in the session offset_, with an offset string suffix reflecting that, while `timestamp (without time zone)` is rendered as-is, with no offset. Additionally, the session offset is used if an offset is not specified when parsing a `with time zone` timestamp from a string.

## Problems
Subtlety arises when casting between these or to and from other types, comparing them, etc e.g. casting a `timestamp with time zone` to a `date` returns different values based on the session offset while casting a `timestamp` to a `date` does not.

Assuming that most production applications support users in multiple time zones and thus localize their times in a rendering and presentation layer (based only on anecdotal experience) rather than at the database connection layer, it seems that the the benefit of supporting non-zero session offsets may not justify the effort required to maintain them code and documentation complexity at this time.

**Comment by dt** [comment]
(assigning a grab bag of people who might have thoughts on this, feel free to remove yourself if you're not interested)

**Comment by mjibson** [comment]
`timestamp` also uses the session offset if there's no zone specified.

**Comment by mjibson** [comment]
That is, parsing is exactly the same for both with and without zone: it uses the session if no zone specified in the string. The only difference is in rendering.

**Comment by dt** [comment]
@mjibson Er, that was a bug -- that meant that they didn't round-trip in the presence of non-zero session offsets -- and was recently fixed.

**Comment by vivekmenezes** [comment]
I think there are two separate issues here:
1. Do we support two different types (timestamp and timestamptz). I'm in favor of only supporting 1.
2. Do we support the session time zone and what are the semantics of that. I think it is okay to support the session time zone as long as the semantics are easy to understand.

**Comment by mjibson** [comment]
Ah, yes. I tested with an old build.

**Comment by mjibson** [comment]
This means any organization that needs to render times in non-UTC will have to do that work in the application. This could get expensive if an organization wants to count the number of daily visitors of a location in that location's timezone. They would have to get back the full timestamp of each row, convert it to whatever timezone, then extract the date. With session offset support they could set the offset, run the extract date function, group and count, and only return one row per day.

Based on the above I'd vote to reject that proposal. We are coming to the point now where we understand exactly what session time zone is and should be used for, and I think we can now document it simply.
